plugins/web/opentelemetry-instrumentation-page-view/src/instrumentation.ts

```diff
@@ -0,0 +1,174 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { InstrumentationBase, isWrapped } from '@opentelemetry/instrumentation';
+import { Event, events, EventLogger } from '@opentelemetry/api-events';
+import { PACKAGE_NAME, PACKAGE_VERSION } from './version';
+import {
+  ApplyCustomEventDataFunction,
+  PageViewInstrumentationConfig,
+} from './types';
+import { PageTypes } from './enums/PageTypes';
+/**
+ * This class represents a page view instrumentation plugin
+ */
+const EVENT_NAME = 'browser.page_view';
+export class PageViewInstrumentation extends InstrumentationBase<PageViewInstrumentationConfig> {
+  eventLogger: EventLogger | null = null;
+  oldUrl = location.href;
+  applyCustomEventData: ApplyCustomEventDataFunction | undefined = undefined;
+
+  /**
+   *
+   * @param config
+   */
+  constructor(config: PageViewInstrumentationConfig) {
+    super(PACKAGE_NAME, PACKAGE_VERSION, config);
+    this.eventLogger = events.getEventLogger(PACKAGE_NAME, PACKAGE_VERSION);
+    this.applyCustomEventData = config?.applyCustomEventData;
+  }
+
+  init() {}
+
+  /**
+   * callback to be executed when using hard navigation
+   */
+  private _onPageView() {
+    const pageViewEvent: Event = {
+      name: EVENT_NAME,
+      data: {
+        url: document.documentURI as string,
+        referrer: document.referrer,
+        title: document.title,
+        type: PageTypes.BASE_PAGE,
+      },
+    };
+    this._applyCustomEventData(pageViewEvent, this.applyCustomEventData);
+    this.eventLogger?.emit(pageViewEvent);
+  }
+
+  /**
+   * callback to be executed when using soft navigation
+   */
+  private _onVirtualPageView(changeState: string | null | undefined) {
+    const title = document.title;
+    const referrer = this.oldUrl;
+    // Don't emit an event if the route didn't change
+    if (referrer === location.href) {
+      return;
+    }
+    const vPageViewEvent: Event = {
+      name: EVENT_NAME,
+      data: {
+        url: window.location.href,
+        title,
+        changeState: changeState || '',
+        referrer,
+        type: PageTypes.VIRTUAL_PAGE,
+      },
+    };
+    this._applyCustomEventData(vPageViewEvent, this.applyCustomEventData);
+    this.eventLogger?.emit(vPageViewEvent);
+  }
+
+  public _seteventLogger(eventLogger: EventLogger) {
+    this.eventLogger = eventLogger;
+  }
+
+  /**
+   * executes callback {_onDOMContentLoaded } when the page is viewed
+   */
+  private _waitForPageLoad() {
+    document.addEventListener('DOMContentLoaded', this._onPageView.bind(this));
```

Comment:
Suggestion: it might make sense to use a buffered PerformanceObserver to listen for PerformanceNavigationTimings, insead of DOMEventListener. I have seen users be unable to find their PageViews because eventListener needs registered at runtime before DOMContentLoaded in order to scrape any data. This problem will also be worse for those have want to install this package via CDN

Reply:
Might be a blocking issue honestly